src/Orleans/Core/GrainFactory.cs

```diff
@@ -37,10 +25,28 @@ public class GrainFactory : IGrainFactory
         private readonly ConcurrentDictionary<Type, IGrainMethodInvoker> invokers =
             new ConcurrentDictionary<Type, IGrainMethodInvoker>();
 
+        /// <summary>
+        /// The cache of typed system target references.
+        /// </summary>
+        private readonly Dictionary<Tuple<GrainId, Type>, Dictionary<SiloAddress, ISystemTarget>> typedSystemTargetReferenceCache =
+                    new Dictionary<Tuple<GrainId, Type>, Dictionary<SiloAddress, ISystemTarget>>();
+
+        /// <summary>
+        /// The cache of type metadata.
+        /// </summary>
+        private readonly TypeMetadataCache typeCache;
+
+        /// <summary>
+        /// The runtime client.
+        /// </summary>
+        private readonly IRuntimeClient runtimeClient;
+
         // Make this internal so that client code is forced to access the IGrainFactory using the 
         // GrainClient (to make sure they don't forget to initialize the client).
-        internal GrainFactory()
+        public GrainFactory(IRuntimeClient runtimeClient, TypeMetadataCache typeCache)
```

Comment:
I agree that it's not the nicest. `IRuntimeClient` is an internal abstraction, but `IGrainFactory` is an external abstraction which exposes some of `IRuntimeClient`s methods. The issue is that OutsideRuntimeClient provides the implementation of CreateObjectReference & DeleteObjectReference. It holds a mapping of id -> obj internally and uses that mapping in those methods and when dispatching calls to those objects.

I could extract all of those methods into some kind of `ILocalObjectManager` or something. It would take a dep on `IMessageCenter` & logger, but I believe that's it. That way, `OutsideRuntimeClient` & `GrainFactory` would depend on `ILocalObjectManager`. On Silos, we could/should inject an implementation which throws for all methods (currently, InsideRuntimeClient throws InvalidOperationException).

If you'd like I can open a separate issue for that?